revocation/internal/crl/crl.go

```diff
@@ -188,38 +279,73 @@ func validate(crl *x509.RevocationList, issuer *x509.Certificate) error {
 	}
 
-	for _, revocationEntry := range baseCRL.RevokedCertificateEntries {
-		if revocationEntry.SerialNumber.Cmp(cert.SerialNumber) == 0 {
-			extensions, err := parseEntryExtensions(revocationEntry)
-			if err != nil {
-				return nil, err
-			}
+	revocationListBundle := []*[]x509.RevocationListEntry{&b.BaseCRL.RevokedCertificateEntries}
+	if b.DeltaCRL != nil {
+		revocationListBundle = append(revocationListBundle, &b.DeltaCRL.RevokedCertificateEntries)
+	}
 
-			// validate signingTime and invalidityDate
-			if !signingTime.IsZero() && !extensions.invalidityDate.IsZero() &&
-				signingTime.Before(extensions.invalidityDate) {
-				// signing time is before the invalidity date which means the
-				// certificate is not revoked at the time of signing.
-				break
+	// latestTempRevokedEntry contains the most recent revocation entry with
+	// reasons such as CertificateHold or RemoveFromCRL.
+	//
+	// If the certificate is revoked with CertificateHold, it is temporarily
+	// revoked. If the certificate is shown in the CRL with RemoveFromCRL,
+	// it is unrevoked.
+	var latestTempRevokedEntry *x509.RevocationListEntry
+
+	// iterate over all the entries in the base and delta CRLs
+	for _, revocationList := range revocationListBundle {
+		for i, revocationEntry := range *revocationList {
+			if revocationEntry.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+				extensions, err := parseEntryExtensions(revocationEntry)
+				if err != nil {
+					return nil, err
+				}
+
+				// validate signingTime and invalidityDate
+				if !signingTime.IsZero() && !extensions.invalidityDate.IsZero() &&
+					signingTime.Before(extensions.invalidityDate) {
+					// signing time is before the invalidity date which means the
+					// certificate is not revoked at the time of signing.
+					break
+				}
+
+				switch revocationEntry.ReasonCode {
+				case int(reasonCodeCertificateHold), int(reasonCodeRemoveFromCRL):
+					// temporarily revoked or unrevoked
+					if latestTempRevokedEntry == nil || latestTempRevokedEntry.RevocationTime.Before(revocationEntry.RevocationTime) {
+						// the revocation status depends on the most recent reason
+						latestTempRevokedEntry = &(*revocationList)[i]
+					}
+				default:
+					// permanently revoked
+					return &result.ServerResult{
+						Result:           result.ResultRevoked,
+						Server:           crlURL,
+						RevocationMethod: result.RevocationMethodCRL,
+					}, nil
+				}
 			}
-
-			// revoked
-			return &result.ServerResult{
-				Result:           result.ResultRevoked,
-				Server:           crlURL,
-				RevocationMethod: result.RevocationMethodCRL,
-			}, nil
 		}
 	}
+	if latestTempRevokedEntry != nil && latestTempRevokedEntry.ReasonCode == reasonCodeCertificateHold {
+		// revoked with CertificateHold
+		return &result.ServerResult{
+			Result:           result.ResultRevoked,
+			Server:           crlURL,
+			RevocationMethod: result.RevocationMethodCRL,
+		}, nil
+	}
```

Comment:
Is this something new and specific to delta CRL?

Reply:
Previously, Pritesh thought we didn't need to handle `reasonCodeCertificateHold` and `reasonCodeRemoveFromCRL` for the base CRL, so we added them for the delta CRL now.